modules/misc/src/materials/ADDensity.C

```diff
@@ -0,0 +1,137 @@
+    params.addCoupledVar("disp_z", "The z displacement");
+    params.addCoupledVar(
+        "displacements",
+        "The displacements appropriate for the simulation geometry and coordinate system");
+    params.addRequiredParam<Real>("density", "Density"););
+
+template <ComputeStage compute_stage>
+ADDensity<compute_stage>::ADDensity(const InputParameters & parameters)
+  : ADMaterial<compute_stage>(parameters),
+    _is_coupled(true),
+    _disp_r(this->template isCoupled("displacements")
+                ? adCoupledValue("displacements", 0)
+                : (this->template isCoupled("disp_r") ? adCoupledValue("disp_r") : adZero())),
+    _orig_density(adGetParam<Real>("density")),
+    _density(adDeclareADProperty<Real>("density"))
+{
+  const bool is_r_coupled = this->template isCoupled("disp_r");
+  const bool is_x_coupled = this->template isCoupled("disp_x");
+  const bool is_y_coupled = this->template isCoupled("disp_y");
+  const bool is_z_coupled = this->template isCoupled("disp_z");
+
+  // new parameter scheme
+  if (this->template isCoupled("displacements"))
+  {
+    // get coordinate system
+    _coord_system = this->template getBlockCoordSystem();
+
+    // get coupled gradients
+    const unsigned int ndisp = this->template coupledComponents("displacements");
+    _grad_disp.resize(ndisp);
+    for (unsigned int i = 0; i < ndisp; ++i)
+      _grad_disp[i] = &adCoupledGradient("displacements", i);
+
+    // fill remaining components with zero
+    _grad_disp.resize(3, &adGradZero());
+  }
+
+  // old deprecated parameters
+  else if (is_x_coupled || is_r_coupled)
+  {
+    // guess(!) coordinate system
+    if (is_r_coupled)
+    {
+      if (is_z_coupled)
+        _coord_system = Moose::COORD_RZ;
+      else
+        _coord_system = Moose::COORD_RSPHERICAL;
+    }
+    else
+      _coord_system = Moose::COORD_XYZ;
+
+    // couple gradients
+    _grad_disp = {is_x_coupled ? &adCoupledGradient("disp_x")
+                         : (is_r_coupled ? &adCoupledGradient("disp_r") : &adGradZero()),
+                  is_y_coupled ? &adCoupledGradient("disp_y")
+                         : (is_z_coupled ? &adCoupledGradient("disp_z") : &adGradZero()),
+                  _coord_system != Moose::COORD_RZ && is_z_coupled ? &adCoupledGradient("disp_z")
+                                                             : &adGradZero()};
+  }
+
+  // no coupling
+  else
+  {
+    _is_coupled = false;
+    // TODO: We should deprecate this case and have the user use a GenericConstantMaterial for this
+  }
+}
+
+template <ComputeStage compute_stage>
+void
+ADDensity<compute_stage>::initQpStatefulProperties()
+{
+  _density[_qp] = _orig_density;
+}
+
+template <ComputeStage compute_stage>
+void
+ADDensity<compute_stage>::computeQpProperties()
+{
+  if (_is_coupled)
+  {
+    // rho * V = rho0 * V0
+    // rho = rho0 * V0 / V
+    // rho = rho0 / det(F)
+    // rho = rho0 / det(grad(u) + 1)
+
+    const auto Axx = (*_grad_disp[0])[_qp](0) + 1.0;
```

Comment:
@lindsayad  Well, using `adZero` and `adGradZero` worked, but now I get a seg fault within these lines during the first linear iteration. It runs initially, the (0) non-linear iteration is calculated `0 Nonlinear |R| = 1.414214e+02`, then it `Access out of bounds in MooseArray (i: 0 size: 0)`. I'm assuming this is due to the complexities added by the AD which adds another vector dimension. This happens when coupling the displacements altogether `displacements = 'disp_x disp_y disp_z'`, or individually. I'm not sure why it works initially, only to have it fail on the first linear iteration. Any ideas?

Reply:
I don't see the reason off the top of my head, but I could probably figure it out pretty quickly if you give me an input file

Reply:
I added tests to `modules/combined/test/tests/elastic_patch`. Interestingly, `ad_elastic_patch_rspherical.i` works, but `ad_elastic_patch_plane_strain.i` and `ad_elastic_patch_rz.i` don't.

Reply:
It's something to do with the differences between RESIDUAL and JACOBIAN template specifications. When I do 
```
    std::cout << "stage: " << compute_stage << "  (*_grad_disp[0])[_qp](0): " << (*_grad_disp[0])[_qp](0) << std::endl;
    std::cout << "stage: " << compute_stage << "  (*_grad_disp[0])[_qp](1): " << (*_grad_disp[0])[_qp](1) << std::endl;
    std::cout << "stage: " << compute_stage << "  (*_grad_disp[0])[_qp](2): " << (*_grad_disp[0])[_qp](2) << std::endl;
    std::cout << "stage: " << compute_stage << "  (*_grad_disp[1])[_qp](0): " << (*_grad_disp[1])[_qp](0) << std::endl;
    std::cout << "stage: " << compute_stage << "  (*_grad_disp[1])[_qp](1): " << (*_grad_disp[1])[_qp](1) << std::endl;
    std::cout << "stage: " << compute_stage << "  (*_grad_disp[1])[_qp](2): " << (*_grad_disp[1])[_qp](2) << std::endl;
```
I get
```
...
stage: 0  (*_grad_disp[0])[_qp](0): 0
stage: 0  (*_grad_disp[0])[_qp](1): 0
stage: 0  (*_grad_disp[0])[_qp](2): 0
stage: 0  (*_grad_disp[1])[_qp](0): 0
stage: 0  (*_grad_disp[1])[_qp](1): 0
stage: 0  (*_grad_disp[1])[_qp](2): 0
 0 Nonlinear |R| = 2.012434e+01
stage: 1  (*_grad_disp[0])[_qp](0): (0,{-3.28615,3.72641,0.88052,-1.32078,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0})
stage: 1  (*_grad_disp[0])[_qp](1): (0,{-32.8615,-12.7359,8.8052,36.7922,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0})
stage: 1  (*_grad_disp[0])[_qp](2): (0,{0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0})
stage: 1  (*_grad_disp[1])[_qp](0): (Segmentation fault: 11
```

Reply:
I can get RESIDUAL to run okay and match with tests. It has to do with the grad terms in the jacobian calculations... help! :)

Reply:
@lindsayad Yeah, something to do with the jacobian compute stage again, but only in RZ. You can reproduce the error with `modules/combined/test/tests/elastic_patch/ad_elastic_patch_rz_nonlinear.i`

Reply:
Should do it!